ddl/ddl_worker_test.go

```diff
@@ -218,22 +221,179 @@ func testCheckJobCancelled(c *C, d *ddl, job *model.Job) {
 		t := meta.NewMeta(txn)
 		historyJob, err := t.GetHistoryDDLJob(job.ID)
 		c.Assert(err, IsNil)
-		c.Assert(historyJob.State, Equals, model.JobCancelled)
+		c.Assert(historyJob.IsCancelled(), IsTrue, Commentf("histroy job %s", historyJob))
 		return nil
 	})
 }
 
 func doDDLJobErr(c *C, schemaID, tableID int64, tp model.ActionType, args []interface{},
 	ctx context.Context, d *ddl) *model.Job {
 	job := &model.Job{
-		SchemaID: schemaID,
-		TableID:  tableID,
-		Type:     tp,
-		Args:     args,
+		SchemaID:   schemaID,
+		TableID:    tableID,
+		Type:       tp,
+		Args:       args,
+		BinlogInfo: &model.HistoryInfo{},
 	}
 	err := d.doDDLJob(ctx, job)
+	// TODO: Add the detail error check.
 	c.Assert(err, NotNil)
 	testCheckJobCancelled(c, d, job)
 
 	return job
 }
+
+func checkCancelState(txn kv.Transaction, job *model.Job, test *testCancelJob) error {
+	var checkErr error
+	if test.cancelState == job.SchemaState {
+		if test.act == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer == 0 {
+		} else {
+			errs, err := inspectkv.CancelJobs(txn, test.jobIDs)
+			if err != test.cancelRetErr {
+				checkErr = errors.Trace(err)
+			}
+			// It only tests cancel one DDL job.
+			if errs[0] != test.cancelRetErrs[0] {
+				checkErr = errors.Trace(errs[0])
+			}
+		}
+	}
+	if test.nextState == job.SchemaState {
+		if test.ddlRetErr != job.Error {
+			checkErr = errors.Trace(job.Error)
+		}
+	}
+	return checkErr
+}
+
+type testCancelJob struct {
+	act           model.ActionType // act is the job action.
+	jobIDs        []int64
+	cancelRetErrs []error // cancelRetErrs is the first return value of CancelJobs.
+	cancelRetErr  error   // cancelRetErr is the second return value of CancelJobs.
+	cancelState   model.SchemaState
+	nextState     model.SchemaState
+	ddlRetErr     error
+}
+
+func buildCancelJobTests(firstID int64) []testCancelJob {
+	nonexist := model.SchemaState(255)
+	err := errCancelledDDLJob
+	errs := []error{err}
+	noErrs := []error{nil}
+	tests := []testCancelJob{
+		{act: model.ActionCreateTable, jobIDs: []int64{firstID}, cancelRetErr: err, cancelRetErrs: noErrs, cancelState: model.StateDeleteOnly, nextState: model.StateWriteOnly, ddlRetErr: err},
+
+		{act: model.ActionAddIndex, jobIDs: []int64{firstID + 1}, cancelRetErrs: errs, cancelState: model.StateDeleteOnly, nextState: model.StateWriteOnly, ddlRetErr: err},
+		{act: model.ActionAddIndex, jobIDs: []int64{firstID + 2}, cancelRetErrs: errs, cancelState: model.StateWriteOnly, nextState: model.StateWriteReorganization, ddlRetErr: err},
+		{act: model.ActionAddIndex, jobIDs: []int64{firstID + 3}, cancelRetErrs: errs, cancelState: model.StateWriteReorganization, nextState: model.StatePublic, ddlRetErr: err},
```

Comment:
If the add index is converted to rollback job, the nex schema state should be StateDeleteOnly.